src/MemStore.cc

```diff
@@ -330,13 +330,16 @@ MemStore::get(const cache_key *key)
     const bool copied = copyFromShm(*e, index, *slot);
 
     if (copied) {
-        e->hashInsert(key);
-        return e;
+        e->hashInsert(cacheKey.key);
+        if (e->preparePublicEntry(cacheKey, true))
```

Comment:
Why do we need to preparePublicEntry() here? There are two possible cases to consider AFAICT:

1. copyFromShm() copied the entire completed memory entry. We are neither a store reader nor a writer at this point and, hence, per [earlier agreement](https://github.com/squid-cache/squid/pull/46#discussion_r141682972), do not need Transients (only writers need Transient entries).
1. copyFromShm() copied a part of the being-written memory entry. We are a store reader and, hence, per [earlier agreement](https://github.com/squid-cache/squid/pull/46#discussion_r141682972), do not need Transients. Please note that, in this partial reading case, there must be a Transients entry already, created by the writer of the entry we are reading. Thus, we may never actually reach MemStore::get() because Controller::find() will call transients->get() instead of calling MemStore::get(). I do not fully understand how this works for non-collapsed MemStore readers because because we seem to need something like Controller::anchorCollapsed() for them, but that method will assert when we are not dealing with a collapsed request. There may be another bug lurking here, but this "do not create Transient entries for readers" comment would still apply even if there is another bug.

Reply:
> There are two possible cases to consider 

How about another case to consider: the entry was loaded from memory cache and then, at some moment, swapping out started? We became a store writer, right? Another example: there was a disk-cached entry, we called SwapDir::get() and Store::Root().memoryOut(), starting copying the entry into the shared memory. Who are we then, reader or writer?

Reply:
N.B. Your other cases do not seem to be relevant to the MemStore::get() method being discussed. I will analyze them in this comment thread, but we are in danger of losing the thread scope here.

> the entry was loaded from memory cache and then, at some moment, swapping out started? We became a store writer, right?

The job (or an equivalent code) swapping out the entry is a writer. That job is responsible for maintaining the "every writer must have a Transients entry" invariant.

> [We were a reader and then] Store::Root().memoryOut(), starting copying the entry into the shared memory. Who are we then, reader or writer?

The same store entry and even the same StoreEntry object can be used by reader(s) and a writer. The writer is responsible for maintaining the "every writer must have a Transients entry" invariant.

In summary, I do not see the need to consider any complex combinations of actions. The problem seems much simpler to me: To become a writer, the job must create the corresponding Transients entry (among other steps that prepare the store entry for writing). For the purpose of this discussion, events that happen before or after that are irrelevant.

The only complication that I can imagine in this scope is a StoreEntry object that needs to be written into two different stores (memory and disk) at the same time. I assume that such concurrent writers either cannot exist in the current code or that they can be easily handled. If my assumption is wrong, we may be able to adjust the code to make it right.

Reply:
> We are a store reader and, hence, per earlier agreement, do not need Transients.
> "every writer must have a Transients entry" invariant.

This invariant looks incomplete because it says nothing about readers. Could you please clarify the wording? From above comments I can conclude something like: "... and reader(which is not attached to writer entry) do_not_need/or_must_not_have Transients entry". 
However, I see problems this such wording, because for collapsing mode we create Transients entry for StoreEntry, before Store writing starts (in fact, it could not start at all, if server fails to respond).

Reply:
General statement: I hesitate attempting to describe all possible relationships involving Transients (and some impossible ones) in a single invariant. Clearly, many invariants can peacefully co-exist, and it may be better to have a few simple invariants than a single complex one. Said that, I am happy to discuss any improvements to the writers invariant -- you can use any tools/approaches/invariants that help you drive this PR to closure, of course!

AFAICT, the invariant was [postulated](https://github.com/squid-cache/squid/pull/46#discussion_r141666456) as

* Every Store entry being written must have a corresponding Transients entry

BTW, the same paragraph also said: A store reader must still attach to the existing Transients entry (if any), of course, but it should not try to create that Transients entry. Thus, the readers were also covered.

FWIW, if I were to polish the writers invariant further, I would probably say something like:

* In SMP mode, every sharable store entry open of writing must have a corresponding Transients entry.

This version still does not detail everything and does not cover cases that lack a sharable store entries open of writing.

> I can conclude something like: "... and reader(which is not attached to writer entry) do_not_need/or_must_not_have Transients entry". However, I see problems this such wording, for collapsing mode we create Transients entry for StoreEntry, before Store writing starts (in fact, it could not start at all, if server fails to respond).

An allowCollapsing() caller is essentially a Store writer. Yes, that caller does not have a store entry or even the content to write (yet), but it creates (i.e., "writes") the public Store metadata that immediately enables cache hits for that future entry/content. That Store metadata consists of the Transients entry and a local StoreEntry object with a public key:

* Successful allowCollapsing(e) creates the Transients entry for `e` and makes `e` public.